### ISSUE TITLE
TDVT generates incorrect SQL with Hive12Dialect

### ISSUE THREAD

**Issue by cornmonster** [tableau/connector-plugin-sdk#189]
**About You:**
Name: Jon
Company: Max compute

**Your question:**
Hi.

Thanks for providing Hive12Dialect! But with this dialect, it seems some generated SQL statement is not correct.

Here is the dialect we are using:
```
<?xml version="1.0" encoding="utf-8"?>
<dialect name='MaxCompute'
         class='maxcompute_jdbc'
         base='Hive12Dialect'
         version='18.1'>
  <function-map>
      <date-function name='DATENAME' return-type='str'>
      <formula part='year'>CAST(YEAR(%2) AS STRING)</formula>
      <formula part='quarter'>CAST(CAST((MONTH(%2) - 1) / 3 + 1 AS BIGINT) AS STRING)</formula>
      <formula part='month'>CAST(MONTH(%2) AS STRING)</formula>
      <formula part='dayofyear'>CAST(DATEDIFF(TO_DATE(%2), TO_DATE(CAST(TRUNC(%2,&apos;YY&apos;) AS DATE))) + 1 AS STRING)</formula>
      <formula part='day'>CAST(DAY(%2) AS STRING)</formula>
      <formula part='weekday'>CAST((8 + DATEDIFF(CAST(%2 AS DATE),NEXT_DAY(%2,&apos;SU&apos;))) AS STRING)</formula>
      <formula part='week'>CAST(FLOOR((14 + DATEDIFF(%2, TRUNC(%2,&apos;YY&apos;)) + DATEDIFF(TRUNC(%2,&apos;YY&apos;),NEXT_DAY(TRUNC(%2,&apos;YY&apos;),&apos;SU&apos;)))/7) AS STRING)</formula>
      <formula part='hour'>CAST(HOUR(%2) AS STRING)</formula>
      <formula part='minute'>COALESCE(CAST(MINUTE(%2) AS STRING), &apos;&apos;)</formula>
      <formula part='second'>COALESCE(CAST(SECOND(%2) AS STRING), &apos;&apos;)</formula>
      <argument type='localstr' />
      <argument type='datetime' />
    </date-function>
    <date-function name='DATENAME' return-type='str'>
      <formula part='year'>CAST(YEAR(%2) AS STRING)</formula>
      <formula part='quarter'>CAST(CAST((MONTH(%2) - 1) / 3 + 1 AS BIGINT) AS STRING)</formula>
      <formula part='month'>CAST(MONTH(%2) AS STRING)</formula>
      <formula part='dayofyear'>CAST(DATEDIFF(TO_DATE(%2), TO_DATE(CAST(TRUNC(%2,&apos;YY&apos;) AS DATE))) + 1 AS STRING)</formula>
      <formula part='day'>CAST(DAY(%2) AS STRING)</formula>
      <formula part='weekday'>CAST((8 + DATEDIFF(CAST(%2 AS DATE),NEXT_DAY(%2,&apos;SU&apos;))) AS STRING)</formula>
      <formula part='week'>CAST(FLOOR((14 + DATEDIFF(%2, TRUNC(%2,&apos;YY&apos;)) + DATEDIFF(TRUNC(%2,&apos;YY&apos;),NEXT_DAY(TRUNC(%2,&apos;YY&apos;),&apos;SU&apos;)))/7) AS STRING)</formula>
      <formula part='hour'>CAST(HOUR(%2) AS STRING)</formula>
      <formula part='minute'>COALESCE(CAST(MINUTE(%2) AS STRING), &apos;&apos;)</formula>
      <formula part='second'>COALESCE(CAST(SECOND(%2) AS STRING), &apos;&apos;)</formula>
      <argument type='localstr' />
      <argument type='datetime' />
      <argument type='localstr' />
    </date-function>
    </function-map>
</dialect>
```

And for test case date.datename.sow.week, DATENAME('week', [date2], 'monday'),  the generated SQL statement was: 
```
SELECT  AS `temp_test__499182808__0_`
FROM `odpsdemo_dev`.`calcs` `calcs`
```

Also, I noticed that in the hive dialect above, the formula part of the two 'DATENAME' functions are identical even through they have different argument list. It seems that the third argument of the second 'DATENAME' function does not present in the formula part at all.

Thank you very much for your time!

Jon
Max Compute, Alibaba

**Comment by liyunbao** [comment]
Hi Jon,

Thanks for your update. It looks like we need to update the DATENAME in Hive12Dialect. I will fix that and get back to you once I finish the test.

Thanks,
Liyun

**Comment by liyunbao** [comment]
Hi Jon,

Can you try the tdd as below? Please note that the base is updated to "HiveDialect".

Thanks,
Liyun

<?xml version="1.0" encoding="utf-8"?>
<dialect name='MaxCompute'
         class='maxcompute_jdbc'
         base='HiveDialect'
         version='18.1'>
  <function-map>
    <date-function name='DATENAME' return-type='str'>
      <formula part='year'>CAST(YEAR(%2) AS STRING)</formula>
      <formula part='quarter'>CAST(CAST((MONTH(%2) - 1) / 3 + 1 AS BIGINT) AS STRING)</formula>
      <formula part='month'>CAST(MONTH(%2) AS STRING)</formula>
      <formula part='dayofyear'>CAST(DATEDIFF(TO_DATE(%2), TO_DATE(CAST(CONCAT(YEAR(%2),&apos;-01-01 00:00:00&apos;) AS TIMESTAMP))) + 1 AS STRING)</formula>
      <formula part='day'>CAST(DAY(%2) AS STRING)</formula>
      <formula part='weekday'>CAST((1 + PMOD(DATEDIFF(TO_DATE(%2), &apos;1995-01-01 00:00:00&apos;), 7)) AS STRING)</formula>     
      <formula part='week'>CAST(FLOOR((7 + DATEDIFF(%2, CONCAT(CAST(YEAR(%2) AS STRING), &apos;-01-01 00:00:00&apos;)) + pmod(8 + pmod(datediff(CONCAT(CAST(YEAR(%2) AS STRING), &apos;-01-01 00:00:00&apos;), &apos;1995-01-01&apos;), 7) - %3 , 7) ) / 7) AS STRING)</formula>
      <formula part='hour'>CAST(HOUR(%2) AS STRING)</formula>
      <formula part='minute'>COALESCE(CAST(MINUTE(%2) AS STRING), &apos;&apos;)</formula>
      <formula part='second'>COALESCE(CAST(SECOND(%2) AS STRING), &apos;&apos;)</formula>
      <argument type='localstr' />
      <argument type='datetime' />
      <argument type='localstr' />
    </date-function>
    </function-map>
</dialect>

**Comment by cornmonster** [comment]
Hi Liyun,

```
<?xml version="1.0" encoding="utf-8"?>
<dialect name='MaxCompute'
         class='maxcompute_jdbc'
         base='HiveDialect'
         version='18.1'>
  <function-map>
      <date-function name='DATENAME' return-type='str'>
      <formula part='year'>CAST(YEAR(%2) AS STRING)</formula>
      <formula part='quarter'>CAST(CAST((MONTH(%2) - 1) / 3 + 1 AS BIGINT) AS STRING)</formula>
      <formula part='month'>CAST(MONTH(%2) AS STRING)</formula>
      <formula part='dayofyear'>CAST(DATEDIFF(TO_DATE(%2), TO_DATE(CAST(TRUNC(%2,&apos;YY&apos;) AS DATE))) + 1 AS STRING)</formula>
      <formula part='day'>CAST(DAY(%2) AS STRING)</formula>
      <formula part='weekday'>CAST((8 + DATEDIFF(CAST(%2 AS DATE),NEXT_DAY(%2,&apos;SU&apos;))) AS STRING)</formula>
      <formula part='week'>CAST(FLOOR((14 + DATEDIFF(%2, TRUNC(%2,&apos;YY&apos;)) + DATEDIFF(TRUNC(%2,&apos;YY&apos;),NEXT_DAY(TRUNC(%2,&apos;YY&apos;),&apos;SU&apos;)))/7) AS STRING)</formula>
      <formula part='hour'>CAST(HOUR(%2) AS STRING)</formula>
      <formula part='minute'>COALESCE(CAST(MINUTE(%2) AS STRING), &apos;&apos;)</formula>
      <formula part='second'>COALESCE(CAST(SECOND(%2) AS STRING), &apos;&apos;)</formula>
      <argument type='localstr' />
      <argument type='datetime' />
    </date-function>
      <date-function name='DATENAME' return-type='str'>
      <formula part='year'>CAST(YEAR(%2) AS STRING)</formula>
      <formula part='quarter'>CAST(CAST((MONTH(%2) - 1) / 3 + 1 AS BIGINT) AS STRING)</formula>
      <formula part='month'>CAST(MONTH(%2) AS STRING)</formula>
      <formula part='dayofyear'>CAST(DATEDIFF(TO_DATE(%2), TO_DATE(CAST(CONCAT(YEAR(%2),&apos;-01-01 00:00:00&apos;) AS TIMESTAMP))) + 1 AS STRING)</formula>
      <formula part='day'>CAST(DAY(%2) AS STRING)</formula>
      <formula part='weekday'>CAST((1 + PMOD(DATEDIFF(TO_DATE(%2), &apos;1995-01-01 00:00:00&apos;), 7)) AS STRING)</formula>
      <formula part='week'>CAST(FLOOR((7 + DATEDIFF(%2, CONCAT(CAST(YEAR(%2) AS STRING), &apos;-01-01 00:00:00&apos;)) + pmod(8 + pmod(datediff(CONCAT(CAST(YEAR(%2) AS STRING), &apos;-01-01 00:00:00&apos;), &apos;1995-01-01&apos;), 7) - %3 , 7) ) / 7) AS STRING)</formula>
      <formula part='hour'>CAST(HOUR(%2) AS STRING)</formula>
      <formula part='minute'>COALESCE(CAST(MINUTE(%2) AS STRING), &apos;&apos;)</formula>
      <formula part='second'>COALESCE(CAST(SECOND(%2) AS STRING), &apos;&apos;)</formula>
      <argument type='localstr' />
      <argument type='datetime' />
      <argument type='localstr' />
    </date-function>
    </function-map>
</dialect>
```

With the tdd above, the generated SQL is still incorrect.

I have attached the test log and results. Hope that could help. 

[tdvt_actuals_combined.zip](https://github.com/tableau/connector-plugin-sdk/files/3366501/tdvt_actuals_combined.zip)
[tdvt_log_combined.txt](https://github.com/tableau/connector-plugin-sdk/files/3366502/tdvt_log_combined.txt)
[tabquery_logs.zip](https://github.com/tableau/connector-plugin-sdk/files/3366503/tabquery_logs.zip)
[test_results_combined.csv.zip](https://github.com/tableau/connector-plugin-sdk/files/3366504/test_results_combined.csv.zip)

Thank you for your time!
Jon

**Comment by liyunbao** [comment]
Hi Jon,

It seems it does not generate the right SQL for DATENAME with 3 arguments. Can you remove the first part definition of "DATENAME"(which includes 2 arguments), and just keep the 3 arguments "DATENAME" definition in your dialect? Or you can use the dialect I mentioned above. Let me know if that works for you.

Thanks,
Liyun

**Comment by liyunbao** [comment]
Same as#191. This issue is fixed and the fix will be available on Tableau 2019.4 release. Hence close it.
